### PR TITLE
Added Copy/Paste handler for Imports to NatTable

### PR DESCRIPTION
**Pull request by sebHollersbacher**
Added Copy and Paste handler to NatTable to correctly handle imports when copy/pasting from/into cells

### CODE REVIEW

File: plugins/org.eclipse.fordiac.ide.application/src/org/eclipse/fordiac/ide/application/properties/EditUntypedSubappInterfaceDataSection.java [hunk 1]
```diff
@@ -194,4 +203,24 @@ private static boolean isConnected(final VarDeclaration varDeclaration) {
...
+			final ListDataProvider<?> provb = (ListDataProvider<?>) NatTableWidgetFactory.getDataLayer(natTable)
+					.getDataProvider();
+			final SelectionLayer selectionLayer = NatTableWidgetFactory.getSelectionLayer(natTable);
+			if (selectionLayer != null && selectionLayer.getSelectedCells().size() == 1) {
+				final var cell = selectionLayer.getSelectedCells().iterator().next();
+				final var rowObj = provb.getRowObject(cell.getRowIndex());
+				if (cell.getColumnIndex() == 1 && rowObj instanceof final VarDeclaration decl) {
+					final Clipboard clipboard = new Clipboard(Display.getDefault());
+					clipboard.setContents(new String[] { PackageNameHelper.getFullTypeName(decl.getType()) },
+							new Transfer[] { TextTransfer.getInstance() });
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.application/src/org/eclipse/fordiac/ide/application/properties/EditUntypedSubappInterfaceDataSection.java]: I believe this will simply copy the full qualified name instead of adding an import when pasted into a type cell? I would suggest creating a separate transfer that also includes the required imports and then add them in the destination type if possible and necessary.

File: plugins/org.eclipse.fordiac.ide.ui/src/org/eclipse/fordiac/ide/ui/widget/NatTableWidgetFactory.java [hunk 2]
```diff
@@ -173,7 +182,7 @@ public static NatTable createRowNatTable(final Composite parent, final DataLayer
 			public void configureUiBindings(final UiBindingRegistry uiBindingRegistry) {
 				uiBindingRegistry.unregisterKeyBinding(new KeyEventMatcher(SWT.MOD1, 'c'));
 				uiBindingRegistry.unregisterKeyBinding(new KeyEventMatcher(SWT.MOD1, 'x'));
-				uiBindingRegistry.registerKeyBinding(new KeyEventMatcher(SWT.MOD1, 'c'), new CopyDataFromTableAction());
+				uiBindingRegistry.registerKeyBinding(new KeyEventMatcher(SWT.MOD1, 'c'), copyAction);
 				uiBindingRegistry.registerKeyBinding(new KeyEventMatcher(SWT.MOD1, 'v'),
 						new PasteDataIntoTableAction(section));
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.ui/src/org/eclipse/fordiac/ide/ui/widget/NatTableWidgetFactory.java]: I would suggest to definine custom command handlers for the `CopyDataToClipboardCommand` and `PasteDataCommand` instead of overriding the default `CopyDataAction` and `PasteDataAction`.

**Review by mx990 — CHANGES_REQUESTED** [review]

File: plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/CopyDataImportCommandHandler.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+			final ListDataProvider<?> provider = (ListDataProvider<?>) dataLayer.getDataProvider();
+			return Arrays.stream(assembledCopiedDataStructure).flatMap(Arrays::stream)
+					.filter(cell -> colMapper.containsKey(columnProvider.getColumns().get(cell.getColumnIndex())))
+					.flatMap(cell -> {
+						if (provider.getRowObject(cell.getRowIndex()) instanceof final EObject decl) {
+							final List<Import> libImports = ImportHelper.getContainerImports(decl);
+							return libImports.stream().filter(imp -> ImportHelper.matchesImport(
+									PackageNameHelper.getFullTypeName(colMapper
+											.get(columnProvider.getColumns().get(cell.getColumnIndex())).apply(decl)),
+									imp));
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/CopyDataImportCommandHandler.java]: I believe this does not cover the case where a type is implicitly imported (i.e., in the same package) in the source, which means there will not be a corresponding `Import` in the enclosing object. I would suggest to simply transfer a set of `Strings` with the full-qualified type name instead and figuring out what to import when pasting.

File: plugins/org.eclipse.fordiac.ide.ui/src/org/eclipse/fordiac/ide/ui/widget/I4diacNatTableUtil.java [hunk 4]
```diff
@@ -21,6 +21,8 @@ public interface I4diacNatTableUtil {
 
 	void removeEntry(Object entry, CompoundCommand cmd);
 
+	void addImport(Object imports, CompoundCommand cmd);
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.ui/src/org/eclipse/fordiac/ide/ui/widget/I4diacNatTableUtil.java]: I would suggest to move this into a paste handler instead of duplicating the respective code 6 times in each section.

File: plugins/org.eclipse.fordiac.ide.datatypeeditor/src/org/eclipse/fordiac/ide/datatypeeditor/widgets/StructEditingComposite.java [hunk 5]
```diff
@@ -223,6 +228,18 @@ public void addEntry(final Object entry, final boolean isInput, final int index,
...
 	}
 
+	@Override
+	public void addImport(final Object imports, final CompoundCommand cmd) {
+		if (imports instanceof final Import[] importsArray
+				&& EcoreUtil.getRootContainer(getType()) instanceof final LibraryElement libElement) {
+			for (final Import imp : importsArray) {
+				if (!ImportHelper.matchesImports(imp.getImportedNamespace(), ImportHelper.getImports(libElement))) {
+					cmd.add(new AddNewImportCommand(libElement, imp.getImportedNamespace()));
+				}
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.datatypeeditor/src/org/eclipse/fordiac/ide/datatypeeditor/widgets/StructEditingComposite.java]: I believe this also needs to skip unnecessary or conflicting imports. In the latter case, we also need to use the FQN when pasting the type. If the import is possible in the destination but was not possible in the source, we should also use the unqualified name. I suggest performing a similar check as in https://github.com/eclipse-4diac/4diac-ide/blob/819cb6252b9688ab77d2530f7aa3333b54d770a7/plugins/org.eclipse.fordiac.ide.model.ui/src/org/eclipse/fordiac/ide/model/ui/widgets/ImportTypeSelectionProposalProvider.java#L50-L60 and then pasting the appropriate name.

**Review by mx990 — CHANGES_REQUESTED** [review]

File: plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/PasteDataImportFromClipboardCommandHandler.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+import org.eclipse.nebula.widgets.nattable.selection.SelectionLayer;
+import org.eclipse.swt.dnd.Clipboard;
+import org.eclipse.swt.widgets.Display;
+
+public class PasteDataImportFromClipboardCommandHandler extends PasteDataFromClipboardCommandHandler {
+
+	private final SelectionLayer selectionLayer;
+	private final CommandStack cmdStack;
+	private BiFunction<TypeLibrary, String, TypeEntry> typeResolver = (typeLib, name) -> typeLib.getDataTypeLibrary()
+			.getDerivedTypeEntry(name);
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/PasteDataImportFromClipboardCommandHandler.java]: It seems the initial value is always overwritten in the constructor?

File: plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/PasteDataImportFromClipboardCommandHandler.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+		this.typeResolver = typeResolver;
+	}
+
+	@Override
+	protected boolean doCommand(final PasteDataCommand command) {
+		super.doCommand(command);
+		final String[] imports = getClipboardContent();
+		final LibraryElement rootElement = getRootElement();
+
+		if (rootElement != null) {
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/PasteDataImportFromClipboardCommandHandler.java:64]: I suggest to perform this check at the beginning, before already pasting the values.

File: plugins/org.eclipse.fordiac.ide.ui/src/org/eclipse/fordiac/ide/ui/widget/PasteDataIntoTableAction.java [hunk 8]
```diff
@@ -52,45 +42,12 @@ public PasteDataIntoTableAction(final I4diacNatTableUtil section) {
...
 	public void run(final NatTable natTable, final KeyEvent event) {
-		final Object clipboardCellsContents = new Clipboard(Display.getCurrent())
-				.getContents(TextTransfer.getInstance());
-		final Object clipboardElementsContents = org.eclipse.gef.ui.actions.Clipboard.getDefault().getContents();
+		final Object clipboardElementsContents = Clipboard.getDefault().getContents();
 
-		if (clipboardCellsContents != null) {
-			pasteClipboardCellsContents(natTable, clipboardCellsContents);
-		} else if (clipboardElementsContents != null) {
+		if (clipboardElementsContents != null) {
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.ui/src/org/eclipse/fordiac/ide/ui/widget/PasteDataIntoTableAction.java:47]: Is there a reason why you cannot also move this into the `PasteDataFromClipboardCommandHandler` and use the vanilla `PasteDataAction`?

File: plugins/org.eclipse.fordiac.ide.ui/src/org/eclipse/fordiac/ide/ui/widget/PasteDataFromClipboardCommandHandler.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+	private final SelectionLayer selectionLayer;
+
+	public PasteDataFromClipboardCommandHandler(final SelectionLayer selectionLayer) {
+		this.selectionLayer = selectionLayer;
+	}
+
+	@Override
+	protected boolean doCommand(final PasteDataCommand command) {
+		final Clipboard clipboard = new Clipboard(Display.getDefault());
+		final Object clipboardCellsContents = clipboard.getContents(TextTransfer.getInstance());
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.ui/src/org/eclipse/fordiac/ide/ui/widget/PasteDataFromClipboardCommandHandler.java:42]: Please make sure to dispose the clipboard afterwards.

File: plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/helpers/ImportHelper.java [hunk 10]
```diff
@@ -182,7 +182,7 @@ private static boolean matchesImports(final String name, final List<Import> impo
 		return imports.stream().anyMatch(imp -> matchesImport(name, imp));
 	}
 
-	private static boolean matchesImport(final String name, final Import imp) {
+	public static boolean matchesImport(final String name, final Import imp) {
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.model/src/org/eclipse/fordiac/ide/model/helpers/ImportHelper.java]: Is this change still needed?

**Review by mx990 — CHANGES_REQUESTED** [review]

File: plugins/org.eclipse.fordiac.ide.ui/src/org/eclipse/fordiac/ide/ui/widget/PasteDataIntoTableAction.java (hunk 8 above)
Comment (sebHollersbacher) [reply plugins/org.eclipse.fordiac.ide.ui/src/org/eclipse/fordiac/ide/ui/widget/PasteDataIntoTableAction.java:47]: I did it similar to the CopyDataFromTableAction
Also, moving it to the CommandHandler would require the RowHeaderDataProvider and I4diacNatTableUtil to be passed to the constuctor and make a things more confusing.

**Comment by sebHollersbacher** [comment]
I'm currently struggling to add the new CommandHandler to every existing table.
The only way I could come up with is to manually add the CommandHandlers, like it is done now in EditUntypedSubappInterfaceDataSection.
@mx990 maybe you have some ideas to resolve this nicely.

**Comment by mx990** [comment]
> I'm currently struggling to add the new CommandHandler to every existing table. The only way I could come up with is to manually add the CommandHandlers, like it is done now in EditUntypedSubappInterfaceDataSection. @mx990 maybe you have some ideas to resolve this nicely.

I would suggest to wrap the relevant code into a configuration, similar to https://github.com/eclipse-4diac/4diac-ide/blob/1826295d7012a6197a841be3e9b2170cf5927151/plugins/org.eclipse.fordiac.ide.ui/src/org/eclipse/fordiac/ide/ui/widget/NatTableWidgetFactory.java#L505-L520
and then add the configuration to the relevant tables.

File: plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/CopyDataImportCommandHandler.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+					.filter(cell -> colMapper.containsKey(columnProvider.getColumns().get(cell.getColumnIndex())))
+					.map(cell -> {
+						if (provider.getRowObject(cell.getRowIndex()) instanceof final EObject decl) {
+							final LibraryElement element = colMapper
+									.get(columnProvider.getColumns().get(cell.getColumnIndex())).apply(decl);
+							if (element instanceof final DataType dt
+									&& (IecTypes.ElementaryTypes.getAllElementaryType().contains(dt)
+											|| IecTypes.GenericTypes.isAnyType(dt))) {
+								return null;
+							}
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/CopyDataImportCommandHandler.java]: I would suggest to filter whether there is a package name instead to cover types without a package as well, for example using `PackageNameHelper.getPackageName(element).isEmpty()`.

File: plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/CopyDataImportCommandHandler.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+							return PackageNameHelper.getFullTypeName(element);
+						}
+						return null;
+					}).filter(Objects::nonNull).filter(s -> !s.isEmpty()).distinct().toArray(size -> new String[size]);
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/CopyDataImportCommandHandler.java]:
```suggestion
					}).filter(Objects::nonNull).filter(Predicate.not(String::isEmpty)).distinct().toArray(String[]::new);
```

File: plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/DefaultImportCopyPasteLayerConfiguration.java [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+	// TODO: adjust DirectlyDerivedTypeTableColumn, TypedElementTableColumn,
+	// VariableTableColumn
+	private static BiFunction<TypeLibrary, String, TypeEntry> getTypeResolver(final NatTableColumn column) {
+		return switch (column) {
+		case final AttributeTableColumn col -> (typeLib, name) -> {
+			final AttributeTypeEntry attributeEntry = typeLib.getAttributeTypeEntry(name);
+			if (attributeEntry != null) {
+				return attributeEntry;
+			}
+			return typeLib.getDataTypeLibrary().getDerivedTypeEntry(name);
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/DefaultImportCopyPasteLayerConfiguration.java:64]: Would it not make more sense to distinguish between the different columns here, similar to `getColMap` below?

File: plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/DefaultImportCopyPasteLayerConfiguration.java [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+		};
+		case final DirectlyDerivedTypeTableColumn col ->
+			(typeLib, name) -> typeLib.getDataTypeLibrary().getDerivedTypeEntry(name);
+		case final TypedElementTableColumn col ->
+			(typeLib, name) -> typeLib.getDataTypeLibrary().getDerivedTypeEntry(name);
+		case final VarDeclarationTableColumn col ->
+			(typeLib, name) -> typeLib.getDataTypeLibrary().getDerivedTypeEntry(name);
+		case final VariableTableColumn col -> (typeLib, name) -> typeLib.getDataTypeLibrary().getDerivedTypeEntry(name);
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/DefaultImportCopyPasteLayerConfiguration.java]: I believe it would improve readability by defining a common (static) method that returns `typeLib.getDataTypeLibrary().getDerivedTypeEntry(name)` and then using a method reference in each of these clauses.

File: plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/PasteDataImportFromClipboardCommandHandler.java [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+		if (conflicts.isEmpty() || selectionLayer.getSelectionModel().getSelections().isEmpty()) {
+			return content;
+		}
+		final var location = selectionLayer.getSelectionAnchor();
+		final int idx = columnProvider.getColumns().indexOf(columns.get(0));
+
+		for (final String[] row : content) {
+			if (conflicts.containsKey(row[idx - location.getColumnPosition()])) {
+				row[idx - location.getColumnPosition()] = conflicts.get(row[idx - location.getColumnPosition()]);
+			}
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/PasteDataImportFromClipboardCommandHandler.java:92]: Please extract `idx - location.getColumnPosition()` into a common variable.

File: plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/PasteDataImportFromClipboardCommandHandler.java [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+
+	@Override
+	protected boolean doCommand(final PasteDataCommand command) {
+		final LibraryElement rootElement = getRootElement();
+		if (rootElement != null) {
+			Arrays.stream(getClipboardContent()).map(imp -> getImportNamespace(rootElement, imp))
+					.filter(Objects::nonNull)
+					.forEach(namespace -> cmdstk.get().execute(new AddNewImportCommand(rootElement, namespace)));
+		}
+		super.doCommand(command);
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/PasteDataImportFromClipboardCommandHandler.java:71]: I believe you also need to clear the conflicts after you are done?

File: plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/PasteDataImportFromClipboardCommandHandler.java [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+
+	private String getImportNamespace(final LibraryElement rootElement, final String imp) {
+		final TypeEntry resolvedType = ImportHelper.resolveImport(PackageNameHelper.extractPlainTypeName(imp),
+				rootElement, name -> typeResolver.apply(rootElement.getTypeLibrary(), name), name -> null);
+
+		if (resolvedType == null) {
+			return imp;
+		}
+
+		if (resolvedType.getFullTypeName().equals(imp)) {
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/PasteDataImportFromClipboardCommandHandler.java]:
```suggestion
		if (resolvedType.getFullTypeName().equalsIgnoreCase(imp)) {
```

**Review by mx990 — CHANGES_REQUESTED** [review]

File: plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/DefaultImportCopyPasteLayerConfiguration.java (hunk 13 above)
Comment (sebHollersbacher) [reply plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/DefaultImportCopyPasteLayerConfiguration.java:64]: When this is used, we don't have any information about the columns, so it would make things more complicated.
For now, it is also the only exception, so I would leaf it like that and change it only if it is required.

File: plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/PasteDataImportFromClipboardCommandHandler.java [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+	}
+
+	@Override
+	protected boolean doCommand(final PasteDataCommand command) {
+		final LibraryElement rootElement = getRootElement();
+		if (rootElement != null) {
+			final var namespaces = Arrays.stream(getClipboardContent()).map(imp -> getImportNamespace(rootElement, imp))
+					.filter(Objects::nonNull).toList();
+			super.doCommand(command);
+			namespaces.forEach(namespace -> cmdstk.get().execute(new AddNewImportCommand(rootElement, namespace)));
```
Comment (mx990) [inline plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/PasteDataImportFromClipboardCommandHandler.java]: I believe you need to add the imports before pasting the type in the super method for it to be resolved correctly?

**Review by mx990 — CHANGES_REQUESTED** [review]

File: plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/PasteDataImportFromClipboardCommandHandler.java (hunk 18 above)
Comment (sebHollersbacher) [reply plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/PasteDataImportFromClipboardCommandHandler.java]: Yes, you are correct. Thank you for pointing that out
Unfortunately, this introduces the bug, where the selection gets cleared after importing (for the StructEditingComposite).

Comment (mx990) [reply plugins/org.eclipse.fordiac.ide.gef/src/org/eclipse/fordiac/ide/gef/nat/PasteDataImportFromClipboardCommandHandler.java]: My guess would be that the problem is caused by a refresh when adding the import? I would suggest to change the handler to use a `CommandExecutor` instead of the `CommandStack` and add a flag to block the refresh in `StructEditingComposite.executeCommand(Command)`, similar to `AbstractSection.executeCommand(Command)`.

**Review by mx990 — APPROVED** [review]

**Review by azoitl — APPROVED** [review]
